Add Pin/Unpin interface to Go

webmaster128 opened the pull request:
Closes #163

ethanfrey commented on api/lib_test.go:
```diff
@@ -70,6 +70,45 @@ func TestCreateFailsWithBadData(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestPin(t *testing.T) {
```
Is there an error case (besides filesystem corruption) we could test?

How about querying for a different checksum that was not registered and ensuring this provides a proper error message, not a panic.

ethanfrey commented on src/cache.rs:
```diff
@@ -123,15 +124,57 @@ pub extern "C" fn load_wasm(
+fn do_pin(
+    cache: &mut Cache<GoApi, GoStorage, GoQuerier>,
+    checksum: Buffer,
+) -> Result<Nothing, Error> {
```
`Nothing`? Not `()`?

webmaster128 replied:
Yeah this is needed because `handle_c_error` requires result value `T: Into<Vec<u8>>`. It was `T: Default` before, but `Checksum` has no sensible `Default`. I better fix `handle_c_error` by having an implementation for binary types and one implementation for default types.

ethanfrey commented on src/nothing.rs:
```diff
@@ -0,0 +1,30 @@
+/// A "no data" type that works with handle_c_error
+pub struct Nothing;
```
Ah, so you can get encoding for it?
Maybe you can extend this docstring

ethanfrey reviewed (commented):
Looks good.
Can you add a test for the error case in pin/unpin?
Then free to merge

maurolacy commented on src/cache.rs:
```diff
@@ -123,15 +124,57 @@ pub extern "C" fn load_wasm(
+        None => Err(Error::empty_arg(CACHE_ARG)),
+    };
+    let _data = handle_c_error(r, err);
+}
```
I guess this is the way to return a potential error to Go land. Strange pattern.

webmaster128 replied:
cgo seems to look into errno to decide if the call was a success or and error and automatically converts a C function like

```
Buffer ibc_packet_timeout(cache_t *cache,
                          Buffer checksum,
                          Buffer env,
                          Buffer msg,
                          DB db,
                          GoApi api,
                          GoQuerier querier,
                          uint64_t gas_limit,
                          bool print_debug,
                          uint64_t *gas_used,
                          Buffer *err);
```

such that it returns a Go style `value, error` pair:

```go
res, err := C.ibc_packet_timeout(cache.ptr, cs, e, p, db, a, q, cu64(gasLimit), cbool(printDebug), &gasUsed, &errmsg)
```

In the case of an error, we pass the message via a different channel (the out error argument `&errmsg`)

See https://utcc.utoronto.ca/~cks/space/blog/programming/GoCgoErrorReturns

ethanfrey reviewed (approved):
pre-emptively approving but please do add a test for the error case before merge

maurolacy reviewed (commented):
LGMT. Just learning this stuff.

webmaster128 commented:
Alright, 4 new commits since your review:
- `Nothing` type was removed again. We can not use all return types `T: Default` and `T: Into<Vec<u8>>` by having two functions. This cannot be easily combined since `Checksum` has no default and `()` cannot be converted into a vector.
- Fixed some wrong argument names
- Added Pin/Unpin tests

ethanfrey commented on api/lib_test.go:
```diff
@@ -88,6 +88,36 @@ func TestPin(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestPinErrors(t *testing.T) {
```
Thanks to add these tests

ethanfrey commented on src/error/rust.rs:
```diff
@@ -142,10 +147,31 @@ where
+/// Otherwise it returns the default value, writes the error message to `errout` and sets [errno].
+///
+/// [errno]: https://utcc.utoronto.ca/~cks/space/blog/programming/GoCgoErrorReturns
+pub fn handle_c_error_default<T>(result: Result<T, RustError>, errout: Option<&mut Buffer>) -> T
```
It seems a good choice to split into two functions, remove some abstraction but make both cases clearer

ethanfrey commented on src/error/rust.rs:
```diff
@@ -242,4 +268,175 @@ mod tests {
     }
+
+    #[test]
+    fn handle_c_error_binary_works() {
```
nice test set

ethanfrey reviewed (approved):
Thanks for the adjustments.
Looks good to merge
